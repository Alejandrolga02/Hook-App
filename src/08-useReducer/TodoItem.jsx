import React from 'react'
import PropTypes from "prop-types";

import circleCheck from './img/circle-check.svg';
import circleX from './img/circle-x.svg';

export const TodoItem = React.memo(({ todo, deleteItem, toggleItem }) => {
	const handleDeleteItem = () => {
		deleteItem(todo.id);
	}

	const handleToggleItem = () => {
		toggleItem(todo.id);
	}

	return (
			<li className="list-group-item d-flex justify-content-between">
				<span className="align-self-center">{todo.description}</span>
				<div>
					<button className="btn btn-danger" onClick={handleDeleteItem}>Delete</button>
					<button className="btn btn-light" onClick={handleToggleItem}>
						{
							(todo.done)
								? <img src={circleCheck} alt="Done icon" />
								: <img src={circleX} alt="Pending icon" />
						}
					</button>
				</div>
			</li>
	);
});

TodoItem.propTypes = {
	todo: PropTypes.object.isRequired,
	deleteItem: PropTypes.func.isRequired,
	toggleItem: PropTypes.func.isRequired
};
TodoItem.displayName = "TodoItem";
