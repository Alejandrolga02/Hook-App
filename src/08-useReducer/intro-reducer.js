const initialState = [
	{
		id: 1,
		todo: 'Learning React',
		done: false
	}
];

const todoReducer = (state = initialState, action = {}) => {

	if (action.type === '[TODO] add todo') {
		return [...state, action.payload]
	}

	return state;
}

let todos = todoReducer();

const newTodo = {
	id: 2,
	todo: 'Joining a Gym',
	done: false
}

const addTodoAction = {
	type: '[TODO] add todo',
	payload: newTodo
}

todos = todoReducer(todos, addTodoAction)
console.log(todos);