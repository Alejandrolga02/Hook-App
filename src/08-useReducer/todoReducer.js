export const initialState = [
	// {
	// 	id: new Date().getTime(),
	// 	description: "Learning React",
	// 	done: false,
	// },
	// {
	// 	id: new Date().getTime() + 10,
	// 	description: "Joining a Gym",
	// 	done: false,
	// },
];

export const todoReducer = (state = [], action) => {
	switch (action.type) {
		case '[TODO] Add Item':
			return [action.payload , ...state];

		case '[TODO] Delete Item':
			return state.filter(task => task.id !== action.payload);

		case '[TODO] Toggle Item':
			return state.map((task) => {
				if (task.id === action.payload) {
					return {
						...task,
						done: !task.done
					}
				}
				return task;
			});

		default:
			return state;
	}
}