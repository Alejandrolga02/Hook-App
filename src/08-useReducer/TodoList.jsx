import PropTypes from "prop-types";
import { TodoItem } from "./TodoItem";
import { useCallback } from "react";

export const TodoList = ({ todos, dispatch }) => {

	const deleteItem = useCallback(
		(id) => {
			return dispatch({
				id,
				type: '[TODO] Delete Item',
			})
		}
	, []);

	const toggleItem = useCallback(
		(id) => {
			return dispatch({
				id,
				type: '[TODO] Toggle Item',
			})
		}
	, []);

	return (
		<>
			<ul className="list-group">
				{
					todos.map((todo) => {
						return <TodoItem key={todo.id} todo={todo} deleteItem={deleteItem} toggleItem={toggleItem}  />
					})
				}
			</ul>
		</>
	);
};

TodoList.propTypes = {
	todos: PropTypes.array.isRequired,
	dispatch: PropTypes.func.isRequired
};
