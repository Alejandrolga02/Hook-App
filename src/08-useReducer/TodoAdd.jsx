import PropTypes from 'prop-types'
import { useForm } from '../hooks/useForm';

export const TodoAdd = ({addTodo}) => {
	const {onInputChange, onResetForm, description} = useForm({
		description: ''
	});
	
	const handleSubmit = (e) => {
		e.preventDefault();
		
		if (description.length < 3) return;  

		addTodo(description);
		onResetForm();
	}

	return (
		<form onSubmit={handleSubmit}>
			<input name='description' onChange={onInputChange} value={description} type="text" placeholder="What do you want to do?" className="form-control" />
			<button className="btn btn-primary" type="submit">Add</button>
		</form>
	);
};

TodoAdd.propTypes = {
	addTodo: PropTypes.func.isRequired
}