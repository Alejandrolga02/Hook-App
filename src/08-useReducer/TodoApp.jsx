import { useCallback, useEffect, useReducer, useState } from "react";
import { initialState, todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

export const TodoApp = () => {
	const [todos, dispatch] = useReducer(todoReducer, initialState);
	const [pendingTasks, setPendingTasks] = useState(0);

	const addTodo = useCallback(
		(description) => {
			dispatch({
				type: '[TODO] Add Item',
				payload: {
					id: new Date().getTime(),
					done: false,
					description,
				}
			});
		}
	, []);

	useEffect(() => {
		const pendingTasksNumber = todos.reduce((counter, task) => {
			if (!task.done) {
				return counter + 1;
			}
			return counter;
		}, 0);
	
		setPendingTasks(pendingTasksNumber);
	}, [todos])
	
	return (
		<>
			<h1>TodoApp: {todos.length} - <small>Pending: {pendingTasks}</small></h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<TodoList todos={todos} dispatch={dispatch} />
				</div>
				<div className="col-5">
					<h4>Add TODO</h4>
					<hr />
					<TodoAdd addTodo={addTodo} />
				</div>
			</div>
		</>
	);
};
