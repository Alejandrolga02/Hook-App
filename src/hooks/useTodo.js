import { useCallback, useEffect, useReducer, useState } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => {
	return JSON.parse(localStorage.getItem('todos')) || [];
}

export const useTodo = () => {
	const [todos, dispatch] = useReducer(todoReducer, [], init);
	const [pendingTasks, setPendingTasks] = useState(0);

	const addTodo = useCallback(
		(description) => {
			dispatch({
				type: '[TODO] Add Item',
				payload: {
					id: new Date().getTime(),
					done: false,
					description,
				}
			});
		}
	, []);

	const deleteItem = useCallback(
		(id) => {
			return dispatch({
				payload: id,
				type: '[TODO] Delete Item',
			})
		}
	, []);

	const toggleItem = useCallback(
		(id) => {
			return dispatch({
				payload: id,
				type: '[TODO] Toggle Item',
			})
		}
	, []);

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));

		const pendingTasksNumber = todos.reduce((counter, task) => {
			if (!task.done) {
				return counter + 1;
			}
			return counter;
		}, 0);
	
		setPendingTasks(pendingTasksNumber);
	}, [todos]);

	return {
		todos,
		addTodo,
		deleteItem,
		toggleItem,
		pendingTasks
	}
}