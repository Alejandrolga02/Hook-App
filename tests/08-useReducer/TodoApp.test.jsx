import { render } from "@testing-library/react";
import { TodoApp } from "../../src/08-useReducer/TodoApp";
import { useTodo } from "../../src/hooks/useTodo";

jest.mock("../../src/hooks/useTodo");

describe("tests in TodoApp", () => {
	useTodo.mockReturnValue({
		todos: [
			{
				"id": 1682531633586,
				"done": false,
				"description": "Get a job"
			},
			{
				"id": 1682531630837,
				"done": false,
				"description": "Joining a Gym"
			},
			{
				"id": 1682531621087,
				"done": false,
				"description": "Graduate from School"
			},
			{
				"id": 1682531606667,
				"done": true,
				"description": "Learn React"
			}
		],
		todosCount: 4,
		pendingTasks: 3,
		deleteItem: jest.fn(),
		toggleItem: jest.fn(),
		addTodo: jest.fn(),
	});

	test("should show component", () => {
		const { container } = render(<TodoApp />);
		expect(container).toMatchSnapshot();
	});
});
