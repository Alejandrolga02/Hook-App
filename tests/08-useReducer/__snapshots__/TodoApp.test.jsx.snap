// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in TodoApp should show component 1`] = `
<div>
  <h1>
    TodoApp: 
    4
     - 
    <small>
      Pending: 
      3
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center"
          >
            Get a job
          </span>
          <div>
            <button
              aria-label="delete-button"
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              aria-label="toggle-button"
              class="btn btn-light"
            >
              <img
                alt="Pending icon"
                src=""
              />
            </button>
          </div>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center"
          >
            Joining a Gym
          </span>
          <div>
            <button
              aria-label="delete-button"
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              aria-label="toggle-button"
              class="btn btn-light"
            >
              <img
                alt="Pending icon"
                src=""
              />
            </button>
          </div>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center"
          >
            Graduate from School
          </span>
          <div>
            <button
              aria-label="delete-button"
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              aria-label="toggle-button"
              class="btn btn-light"
            >
              <img
                alt="Pending icon"
                src=""
              />
            </button>
          </div>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center"
          >
            Learn React
          </span>
          <div>
            <button
              aria-label="delete-button"
              class="btn btn-danger"
            >
              Delete
            </button>
            <button
              aria-label="toggle-button"
              class="btn btn-light"
            >
              <img
                alt="Done icon"
                src=""
              />
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="description"
          placeholder="What do you want to do?"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
